vtStorAtaProtocol/Platform/Linux/ProtocolAtaPassThrough.cpp

```diff
@@ -27,164 +30,135 @@ namespace vtStor
-        }        
-
-        return(errorCode);
-    }
-*/
-   /* void cAtaPassThrough::InitializePassThroughDirect( const StorageUtility::Ata::sCommandCharacteristic& CommandCharacteristics, const StorageUtility::Ata::uTaskFileRegister& PreviousTaskFile, const StorageUtility::Ata::uTaskFileRegister& CurrentTaskFile, std::shared_ptr<cBufferInterface> DataBuffer, U32 TimeoutValueInSeconds )
+    cBufferFormatter bufferFormatter = cBufferFormatter::Reader(Essense);
+
+    switch (bufferFormatter.GetHeader().Format)
+    {
+    case 1:
     {
-        m_AtaPassThrough.Length = sizeof( ATA_PASS_THROUGH_DIRECT );
-        m_AtaPassThrough.DataTransferLength = CommandCharacteristics.DataTransferLengthInBytes;
-        m_AtaPassThrough.TimeOutValue = TimeoutValueInSeconds;
-        if (nullptr != DataBuffer)
+        cEssenseAta1 essense = cEssenseAta1::Reader(Essense);
+
+        InitializePassThroughDirect(essense.GetCommandCharacteristics(), essense.GetTaskFileExt(), essense.GetTaskFile(), DataBuffer, 60);
+
+        if (IOCTL_SG_IO_ERROR == ioctl(Handle.Handle, SG_IO, &m_ATAPassThrough))
         {
-            m_AtaPassThrough.DataBuffer = DataBuffer->ToDataBuffer();
+            //! ERROR: IOCTL SG_IO was not successful
+            return(eErrorCode::Io);
         }
-        m_AtaPassThrough.ReservedAsUchar = 0;
-        m_AtaPassThrough.ReservedAsUlong = 0;
+    } break;
 
-        InitializeFlags( CommandCharacteristics );
-        InitializeTaskFileInputRegisters( PreviousTaskFile, CurrentTaskFile );
+    default:
+    {
+        errorCode = eErrorCode::FormatNotSupported;
+    } break;
     }
 
-    void cAtaPassThrough::InitializeFlags( const StorageUtility::Ata::sCommandCharacteristic& AtaCommandCharacteristic )
-    {
-        // Clear all flags
-        m_AtaPassThrough.AtaFlags = 0;
+    return(errorCode);
+}
 
-        if (StorageUtility::Ata::eDeviceReadyFlag::DEVICE_READY_REQUIRED == AtaCommandCharacteristic.DeviceReadyFlag)
-        {
-            m_AtaPassThrough.AtaFlags |= ATA_FLAGS_DRDY_REQUIRED;
-        }
+void cAtaPassThrough::InitializePassThroughDirect(
+        const StorageUtility::Ata::sCommandCharacteristic& CommandCharacteristics,
+        const StorageUtility::Ata::uTaskFileRegister& PreviousTaskFile,
+        const StorageUtility::Ata::uTaskFileRegister& CurrentTaskFile,
+        std::shared_ptr<cBufferInterface> DataBuffer,
+        U32 TimeoutValueInSeconds
+        )
+{
+    memset(&m_SenseData, 0, sizeof(m_SenseData));
+    memset(&m_ATAPassThrough, 0, sizeof(sg_io_hdr));
 
-        switch (AtaCommandCharacteristic.DataAccess)
-        {
-        case StorageUtility::Ata::eDataAccess::READ_FROM_DEVICE:
-        {
-            m_AtaPassThrough.AtaFlags |= ATA_FLAGS_DATA_IN;
-        } break;
-        case StorageUtility::Ata::eDataAccess::WRITE_TO_DEVICE:
-        {
-            m_AtaPassThrough.AtaFlags |= ATA_FLAGS_DATA_OUT;
-        } break;
-        }
+    m_ATAPassThrough.interface_id     = 'S';

```

Comment:
It is a option to identify using SCSI generic, we sent Ata command pass throught SCSI on Linux. I will define with int type.